src/combinatorics/graph/basic.lean

```diff
@@ -0,0 +1,620 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import combinatorics.quiver
+
+/-!
+# Graph typeclasses
+
+The hierarchy of graph-like structures
+-/
+
+universes u u₀ u₁ u₂ u₃ u₄
+variables {F α β W : Type*}
+
+/-! ### Prop world -/
+
+-- This is an offer to refactor `rel` as well. It feels like it belongs to this hierarchy and
+-- refactoring it should be fine as it's not used much.
+/-- Relation. -/
+structure rel (α β : Sort*) :=
+(to_fun : α → β → Prop)
+
+/-- `rel_class F α` states that `F` is a type of relations between `α` and `β`.
+
+Extend this class when you extend `rel`. -/
+class rel_class (F α β : Sort*) :=
+(to_fun : F → α → β → Prop)
+
+instance rel.rel_class : rel_class (rel α β) α β :=
+⟨rel.to_fun⟩
+
+/-- The adjacency relation of a graph. -/
+def graph.adj {F α β : Sort*} [rel_class F α β] : F → α → β → Prop := rel_class.to_fun
```

Comment:
Why do we need this if this is exactly `rel_class.to_fun`?

Reply:
`rel_class.to_fun` is not exactly a sexy name!